### PR TITLE
#353 Easier flatExtracting usage

### PR DESCRIPTION
**Pull request by AlexBischof**
Added flatExtracting(String) to AbstractIterableAssert and AbstractObjectArrayAssert and reused overloaded flatExtracting. Interally an extractor is created for the property name.

I've converted the IllegalArgumentException into a NPE so that the behavior of the existing flatExtracting is same to the new one, as you can see from the old tests.

### CODE REVIEW

File: src/main/java/org/assertj/core/api/AbstractIterableAssert.java [hunk 1]
```diff
@@ -924,6 +925,45 @@ public S usingDefaultElementComparator() {
...
+     * homer.getChildren().add(bart);
+     * homer.getChildren().add(lisa);
+     * homer.getChildren().add(maggie);
+     *
+     * CartoonCharacter pebbles = new CartoonCharacter("Pebbles Flintstone");
+     * CartoonCharacter fred = new CartoonCharacter("Fred Flintstone");
+     * fred.getChildren().add(pebbles);
+     *
+     *
+     * assertThat(Lists.newArrayList(homer, fred)).flatExtracting("children").containsOnly(bart, lisa, maggie, pebbles);
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractIterableAssert.java]: get rid of `Lists.` as we use `newArrayList` directly when creating `cartoonCharacters`, it will make the example lighter.
It would also be good to put `.containsOnly` in the next line, aligned on `.flatExtracting`

File: src/main/java/org/assertj/core/api/AbstractIterableAssert.java [hunk 2]
```diff
@@ -924,6 +925,45 @@ public S usingDefaultElementComparator() {
 	return new ListAssert<>(result);
   }
 
+    /**
+     * Extract the Iterable values from Iterable's elements under test by applying an Iterable extracting function based
```
Comment (joel-costigliola) [inline src/main/java/org/assertj/core/api/AbstractIterableAssert.java]: I feel we could simplify this explanation by avoiding to mention the extracting function as it is an implementation detail. Maybe something like:
```
Extract from Iterable's elements the Iterable values corresponding to the given property/field name and concatenate them into a single list becomes a new object under test.
```

File: src/test/java/org/assertj/core/api/objectarray/ObjectArrayAssert_flatExtracting_Test.java [hunk 3]
```diff
@@ -68,10 +68,4 @@ public void should_allow_assertions_on_empty_result_lists() throws Exception {
     assertThat(new CartoonCharacter[] { bart, lisa, maggie }).flatExtracting(children).isEmpty();
   }
 
-  @Test
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/objectarray/ObjectArrayAssert_flatExtracting_Test.java]: why is this test removed ?

File: src/test/java/org/assertj/core/api/objectarray/ObjectArrayAssert_flatExtractingEasy_Test.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  @Test
+  public void should_allow_assertions_on_joined_lists_when_extracting_children() {
+    assertThat(new CartoonCharacter[] { homer, fred }).flatExtracting("children").containsOnly(bart, lisa, maggie,
+        pebbles);
+  }
+
+  @Test
+  public void should_allow_assertions_on_empty_result_lists() throws Exception {
+    assertThat(new CartoonCharacter[] { bart, lisa, maggie }).flatExtracting("children").isEmpty();
+  }
```
Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/objectarray/ObjectArrayAssert_flatExtractingEasy_Test.java]: add test to assess NPE when one of the elements is null

Comment (AlexBischof) [inline src/test/java/org/assertj/core/api/objectarray/ObjectArrayAssert_flatExtractingEasy_Test.java]: The NPE of the older version was related to the extractor implementation in the test class. With the new version an IllegalArgumentException is thrown. My first idea was to rethrow the IllegalArgumentException as a NPE but that doesn't look good. So, just a test for the IllegalArgumentException?

     private final Extractor<CartoonCharacter, List<CartoonCharacter>> children = new Extractor<CartoonCharacter, List<CartoonCharacter>>() {
        @Override
        public List<CartoonCharacter> extract(CartoonCharacter input) {
          return input.getChildren();
        }
      };

Comment (joel-costigliola) [inline src/test/java/org/assertj/core/api/objectarray/ObjectArrayAssert_flatExtractingEasy_Test.java]: Yes ! `IllegalArgumentException` express the behavior ina better way than NPE

**Comment by joel-costigliola** [comment]
`IterableAssert_flatExtractingEasy_test` as now the code throws an `IllegalArgumentException`.
I'm fixing that but please run the unit tests before pushing code.

**Comment by AlexBischof** [comment]
Sorry for that but i always run mvn clean test or install before i push. The problem is/was that the test class has a wrong name for the maven convention (begin or end with Test (casesensitive)). So the test classes should end with 'Test'. Does it makes sense that i rename the classes in this PR or would you like to do it?

**Comment by joel-costigliola** [comment]
You are indeed right, I will rename the class.
